paddle/framework/net_design.md

```diff
@@ -0,0 +1,159 @@
+# Network Design
+
+`Network` is the container and controller of a set of operators in a network, 
+users can build a real network from a `NetDef` in protobuf message 
+and use `Network.Run()` to run all the operators in the network.
+
+The `Network` will
+
+- manage all the operators contained in the network.
+- not own any `Variable`.
+
+# API
+
+## NetworkBase
+To make the `Network` extendable, a base class is defined like this
+
+```c++
+// The minimum a network should be implemented.
+class NetworkBase {
+ public:
+  // `def` is a proto message that describe the structure of a network.
+  NetworkBase(const NetDef& def);
+
```

Comment:
I think `addOperator` should return a Index. Maybe code like this,

```cpp
class NetworkBase {
public:
  using OpIndex = size_t;

  OpIndex addOperator(const OpDesc& op);
};
```

So that `Run` method can take a `begin` and `end` argument like:

```cpp
bool Run(Scope* scope, OpIndex begin = -1UL, OpIndex end = -1UL);
```

It will be very helpful if a network can run partially.